---
title: "RSC_jaredYseq2022_8-22-2022"
author: "Nicole Adams"
date: '2022-08-22'
output: 
  html_document:
    toc: true
    code_folding: show
---

# Genotyping samples sequenced by Jared (2021) with those sequenced in 2022
This work flow is for genotyping and basic SNP filtering of invasive Red Swamp Crayfish (*P. clarkii*) samples sequenced in 2021 by Jared Homola and in 2022 by Nicole Adams for juvenile cohort identification, estimates of breeding adults and reproductive success, and pedigree reconstruction. This is for the manuscript Adams NE, Homola JJ, Sard NM, Nathan LR, Roth BM, Robinson JD, Scribner KT. 2024. Genomic data characterize reproductive ecology patterns in Michigan invasive Red Swamp Crayfish (*Procambarus clarkii*). *Evolutionary Applications*. 

&nbsp;

The preceding bioinformatic processing for these files can be found in RSC_processing2022_4ms.Rmd and RSC_processingNickJared_4ms.Rmd. The original document this Rmd is based on is RSC_jaredYseq2022_8-22-22.Rmd.

&nbsp;

#### Load R libraries
```{r, message=FALSE, warning=FALSE}
library(tidyverse)
library(ggrepel)

#BiocManager::install("SeqArray")
library(SeqArray)
library(SNPRelate)
library(vcfR)
```

&nbsp;

## Combine Jared and seq2022 sample files
combine pop files
```{bash, eval=FALSE}
cd /mnt/home/adamsn23/RedSwampCrayfish/jaredYseq2022/SHELL/dependencies
touch SHELL/dependencies/popMap_jaredYseq2022.txt

cat SHELL/dependencies/popMap_JH_nam.txt > SHELL/dependencies/popMap_jaredYseq2022.txt 
cat SHELL/dependencies/popMap_seq2022.txt >> SHELL/dependencies/popMap_jaredYseq2022.txt 
# N=3411

awk -v OFS='\t' '{ print $1, $2 }' popMap_jaredYseq2022.txt > popMap_jaredYseq2022.2.txt
```

&nbsp;

### Copy filtered BAMs into one directory
```{bash, eval=FALSE}
 
cp FC5*.pe.sorted.filtered.bam $SCRATCH/RSC/jaredYseq2022/bams
cp FC1*.pe.sorted.filtered.bam $SCRATCH/RSC/jaredYseq2022/bams
cp FC1*.pe.sorted.filtered.bam $SCRATCH/RSC/jaredYseq2022/bams
FC8

cd /mnt/home/adamsn23/RedSwampCrayfish/nickYjaredReanalysis/OUT/filtered
cp *.pe.sorted.filtered.bam $SCRATCH/RSC/jaredYseq2022/bams

cp Cfod*.pe.sorted.filtered.bam /mnt/gs18/scratch/users/adamsn23/RSC/jaredYseq2022/bams

```

&nbsp;

## Genotype with gstacks
(/mnt/home/adamsn23/RedSwampCrayfish/jaredYseq2022/scripts/13_genotype_NEA.sh)
```{bash, eval=FALSE}
#==================================================================================================
#   File: genotype.sh
#   Date: 01/07/20, 03/14/2022 (NEA), 07/16/2022 (NEA)
#   Description: Genotype using GStacks
#--------------------------------------------------------------------------------------------------
#       Authors: Jared Homola, Nicole Adams
#==================================================================================================

#Define alias for project root directory
RUN_PATH=/mnt/home/adamsn23/RedSwampCrayfish/jaredYseq2022
SCR=/mnt/gs18/scratch/users/adamsn23/RSC/jaredYseq2022

mkdir $RUN_PATH/OUT/genotyped/

cd $RUN_PATH/SHELL

echo '#!/bin/sh 
#SBATCH --nodes=1-8
#SBATCH --ntasks=1
#SBATCH -t 9:00:00
#SBATCH --cpus-per-task=24
#SBATCH --mem-per-cpu=16G
#SBATCH -J genotype
#SBATCH -o '$RUN_PATH'/QSTAT/genotype.o
#SBATCH --error '$RUN_PATH'/QSTAT/genotype.err

module purge
module load GCC/9.3.0 OpenMPI/4.0.3 Stacks/2.59

cd '$RUN_PATH'/OUT/filtered/
gstacks -I '$SCR'/bams \
    -S .pe.sorted.filtered.bam \
    --threads 24 \
    -M '$RUN_PATH'/SHELL/dependencies/popMap_jaredYseq2022.txt \
    -O '$RUN_PATH'/OUT/genotyped/

scontrol show job ${SLURM_JOB_ID} $' > genotype.sh

sbatch genotype.sh


```

&nbsp;
&nbsp;

## Create VCF with Stacks populations
(/mnt/home/adamsn23/RedSwampCrayfish/jaredYseq2022/scripts/14_populations_NEA.sh)
```{bash, eval=FALSE}
#==================================================================================================
#   File: 14_populations.sh
#   Date: 01/07/20, 03/17/2022 (NEA), 07/17/2022 (NEA)
#   Description: Run Stacks' 'Populations' to generate vcf file
#	Authors: Jared Homola, Nicole Adams
#==================================================================================================

#Define alias for project root directory
RUN_PATH=/mnt/home/adamsn23/RedSwampCrayfish/jaredYseq2022
SCR=/mnt/gs18/scratch/users/adamsn23/RSC/jaredYseq2022

cd $RUN_PATH/SHELL

echo '#!/bin/sh 
#SBATCH --nodes=1-4
#SBATCH --ntasks=1
#SBATCH -t 9:00:00
#SBATCH --cpus-per-task=16
#SBATCH --mem-per-cpu=8G
#SBATCH -J populations
#SBATCH -o '$RUN_PATH'/QSTAT/populations_jY22.o

module purge
module load GCC/9.3.0 OpenMPI/4.0.3 Stacks/2.59


populations -P '$RUN_PATH'/OUT/genotyped \
    -O '$RUN_PATH'/OUT/genotyped \
    -t 16 \
    -M '$RUN_PATH'/SHELL/dependencies/popMap_jaredYseq2022.txt \
    --vcf \
    --fasta-loci \
    --ordered-export

scontrol show job ${SLURM_JOB_ID} $' > populations_jY22.sh

sbatch populations_jY22.sh


```

&nbsp;
&nbsp;

## Filter SNPs
Taken from the first part of 15_snpAnalyses_NEA.sh. First, filter at the genotype level (-min-meanDP 2 --minGQ 20 --mac 3 --max-missing 0.01). Followed by removing individuals with >99% missing data.
(/mnt/home/adamsn23/RedSwampCrayfish/jaredYseq2022/scripts/filterSNPs.sbatch) 
```{bash, eval=FALSE}
#!/bin/sh 
#SBATCH --nodes=1-4
#SBATCH --ntasks=8
#SBATCH -t 3:00:00
#SBATCH --cpus-per-task=4
#SBATCH --mem-per-cpu=8G
#SBATCH -J Count_reads
#SBATCH -o /mnt/home/adamsn23/RedSwampCrayfish/jaredYseq2022/QSTAT/filterSNPs.o
#SBATCH --error /mnt/home/adamsn23/RedSwampCrayfish/jaredYseq2022/QSTAT/filterSNPs.err

#==================================================================================================
#   File: 15_snpAnalyses.sh
#   Date: 01/07/20, 07/17/2022 (NEA)
#   Description: Analyze retained SNPs
#--------------------------------------------------------------------------------------------------
#       Authors: Jared Homola, Nicole Adams
#==================================================================================================

# Make directory and load modules
RUN_PATH=/mnt/home/adamsn23/RedSwampCrayfish/jaredYseq2022
SCR=/mnt/gs21/scratch/adamsn23

cd $RUN_PATH/OUT/genotyped
module load GCC/7.3.0-2.30 OpenMPI/3.1.1 VCFtools/0.1.15-Perl-5.28.0

### Initial number of SNPs: 
egrep -v "^#" populations.snps.vcf | wc -l 

### 1st round of filtering: Filter at genotype level (-min-meanDP 2 --minGQ 20 --mac 3 --max-missing 0.01)
vcftools --vcf populations.snps.vcf --min-meanDP 2 --minGQ 20 --mac 3 --max-missing 0.01 --recode --recode-INFO-all --stdout > filteredSNPs_jY22_8-1.vcf # 92,634 loci 
egrep -v "^#" filteredSNPs_jY22_8-1.vcf | wc -l #92634

## 2nd round of filtering: rm individuals with >99% missing data
vcftools --vcf filteredSNPs_jY22_8-1.vcf --missing-indv --out missingIndv_jY22_8-1 
cat missingIndv_jY22_8-1.imiss | awk '$5 < 0.99' | awk '{print $1}' > good_individuals_jY22_8-1.list 

vcftools --vcf filteredSNPs_jY22_8-1.vcf --keep good_individuals_jY22_8-1.list --recode --recode-INFO-all --stdout > filteredSNPsTMP_jY22_8-1.vcf  

cat good_individuals_jY22_8-1.list |wc -l #3408

```

Initial 2 rounds of filtering resulted in 3,408 individuals and 92,634 SNPs

&nbsp;

### Iterative filtering
18 rounds of iterative filtering (/mnt/home/adamsn23/RedSwampCrayfish/jaredYseq2022/scripts/iterativeFilterSNPs.sh)
```{bash, eval=FALSE}
#!/bin/sh 
#SBATCH --nodes=1-4
#SBATCH --ntasks=8
#SBATCH -t 6:00:00
#SBATCH --cpus-per-task=4
#SBATCH --mem-per-cpu=8G
#SBATCH -J filter.snps
#SBATCH -o /mnt/home/adamsn23/RedSwampCrayfish/jaredYseq2022/QSTAT/ifilterSNPs.o
#SBATCH --error /mnt/home/adamsn23/RedSwampCrayfish/jaredYseq2022/QSTAT/ifilterSNPs.err



################ SNP filtering for pop gen analyses ################
cd /mnt/home/adamsn23/RedSwampCrayfish/jaredYseq2022/OUT/genotyped
module load GCC/7.3.0-2.30 OpenMPI/3.1.1 VCFtools/0.1.15-Perl-5.28.0

rounds=10
#maxmissing_val=0.0375
#missingIndv_val=0.975
maxmissing_val=0.3375
missingIndv_val=0.775


### First iteration done manually to remove the very worst performing loci and individuals and generate "filteredSNPs.vcf" as input
#vcftools --vcf populations.snps.vcf --min-meanDP 2 --minGQ 20 --mac 3 --max-missing 0.01 --recode --recode-INFO-all --stdout > filteredSNPs_jY22_8-1.vcf
#vcftools --vcf filteredSNPs_jY22_8-1.vcf --missing-indv --out missingIndv_jY22_8-1
#cat missingIndv_jY22_8-1.imiss | awk '$5 < 0.99' | awk '{print $1}' > good_individuals_jY22_8-1.list
#vcftools --vcf filteredSNPs_jY22_8-1.vcf --keep good_individuals_jY22_8-1.list --recode --recode-INFO-all --stdout > filteredSNPsTMP_jY22_8-1.vcf

#less -S good_individuals_jY22_8-1.list | wc -l
#grep -v "#" filteredSNPsTMP_jY22_8-1.vcf | wc -l
### Above left 92,634 loci and 3,408 individuals with min-meanDP 2 and max-missing 0.01

currentRound=1
until [ $currentRound -gt $rounds ]
do
    echo Round: $currentRound maxmissing_val: $maxmissing_val missindIndv_val: $missingIndv_val

    vcftools --vcf filteredSNPsTMP_jY22_8-1.vcf --minDP 7 --minGQ 20 --mac 3 --max-missing $maxmissing_val --recode --recode-INFO-all --stdout > filteredSNPsTMP1_jY22_8-2.vcf
    vcftools --vcf filteredSNPsTMP1_jY22_8-2.vcf --missing-indv --out missingIndv_jY22_8-2
    cat missingIndv_jY22_8-2.imiss | awk '$5 < '$missingIndv_val'' | awk '{print $1}' > good_individuals_jY22_8-2.list
    vcftools --vcf filteredSNPsTMP1_jY22_8-2.vcf --keep good_individuals_jY22_8-2.list --recode --recode-INFO-all --stdout > filteredSNPsTMP2_jY22_8-2.vcf
  #  mv filteredSNPsTMP2.vcf filteredSNPs.vcf

    
    echo Round: $currentRound maxmissing_val: $maxmissing_val missindIndv_val: $missingIndv_val >> filterResults.txt

    less -S good_individuals_jY22_8-2.list | wc -l >> filterResults.txt
    grep -v "#" filteredSNPsTMP2_jY22_8-2.vcf | wc -l >> filterResults.txt    

    missingIndv_val=$(echo "($missingIndv_val - 0.025)" | bc -l)
    maxmissing_val=$(echo "($maxmissing_val + 0.0375)" | bc -l)
    ((currentRound++))


done

```

&nbsp;

#### Look at iterative filtering results in R
```{r, warning=FALSE, message=FALSE}
max.miss <- seq(0.0375, 0.9375, by=0.0375)
miss.indiv <- seq(0.975, 0.375, by=-0.025)
indiv <- c(3154, 3035, 2957, 2901, 2858, 2829, 2801, 2779, 2756, 2742, 2730, 2721, 2715, 2710, 2708, 2712, 2730, 2746, 2770, 2799, 2843, 2918, 2996, 3098, 0)
sites <- c(80916, 80605, 76878, 71261, 66697, 62681, 59151, 55830, 52753, 49841, 46732, 43795, 40838, 37678, 34619, 31078, 27322, 23283, 19154, 14597, 9734, 5253, 1967, 302, 0)

it.filt <- as.data.frame(cbind(max.miss, miss.indiv, indiv, sites))

it.filt.l <- it.filt %>% pivot_longer(!c(max.miss, miss.indiv), names_to = "resultType", values_to = "count")

max.miss.p <- ggplot(it.filt.l %>% filter(count >0), aes(x=max.miss, y=count, color=resultType)) +
  geom_point(show.legend = F) +
  facet_wrap(~resultType, scales = "free_y") +
  theme(text = element_text(size = 20)) +
  geom_vline(xintercept = 0.75, linetype="dashed", color="gray") +
  theme_minimal()

miss.indiv.p <- ggplot(it.filt.l %>% filter(count >0), aes(x=miss.indiv, y=count, color=resultType)) +
  geom_point(show.legend = F) +
  scale_x_reverse() +
  facet_wrap(~resultType, scales = "free_y") +
  theme(text = element_text(size = 20)) +
  geom_vline(xintercept = 0.50, linetype="dashed", color="gray") +
  theme_minimal()

cowplot::plot_grid(max.miss.p, miss.indiv.p, ncol = 1)  


```

Iterative filtering suggests cutoffs of 0.75 for max missing for genotypes and 0.50 for individual missingness. 

&nbsp;

### Filtering based on iterative results
Filter at the genotype level again (--minDP 7 --minGQ 20 --mac 3 --max-missing 0.75)
```{bash, eval=FALSE}
cd /mnt/home/adamsn23/RedSwampCrayfish/jaredYseq2022/OUT/genotyped
module load GCC/7.3.0-2.30 OpenMPI/3.1.1 VCFtools/0.1.15-Perl-5.28.0

vcftools --vcf filteredSNPsTMP_jY22_8-1.vcf --minDP 7 --minGQ 20 --mac 3 --max-missing 0.75 --recode --recode-INFO-all --stdout > filteredSNPs_jY22_8-1_dp7miss75.vcf

```

&nbsp;

## Further filtering for analyses based on juveniles
Filter individual missingness at 50%
```{bash, eval=FALSE}
cd /mnt/home/adamsn23/RedSwampCrayfish/jaredYseq2022/OUT/genotyped
module load GCC/7.3.0-2.30 OpenMPI/3.1.1 VCFtools/0.1.15-Perl-5.28.0

vcftools --vcf filteredSNPs_jY22_8-1_dp7miss75.vcf --missing-indv --out filteredSNPs_jY22_8-1_dp7miss75
cat filteredSNPs_jY22_8-1_dp7miss75.imiss | awk '$5 < 0.5' | awk '{print $1}' > good_individuals_jY22_8-1_dp7miss75.list #2799

vcftools --vcf filteredSNPs_jY22_8-1_dp7miss75.vcf --keep good_individuals_jY22_8-1_dp7miss75.list --recode --recode-INFO-all --stdout > filteredSNPs_jY22_8-1_dp7miss75miss50.vcf
 
grep -v "#" filteredSNPs_jY22_8-1_dp7miss75miss50.vcf | wc -l #14597

```

&nbsp;

### Remove sites with heterozygosity > 0.6
#### Step 1: Calculate heterozygosity
```{bash, eval=FALSE}
#Take filteredSNPs_jY22_8-1_dp7miss75miss50.vcf and remove SNPs with observed heterozygosity > 0.6 * 
cd /mnt/home/adamsn23/RedSwampCrayfish/jaredYseq2022/OUT/genotyped
module load GCC/7.3.0-2.30 OpenMPI/3.1.1 VCFtools/0.1.15-Perl-5.28.0

#1) calc hwe 
vcftools --vcf filteredSNPs_jY22_8-1_dp7miss75miss50.vcf --hardy --out filteredSNPs_jY22_8-1_dp7miss75miss50.h 

#2) 
cat filteredSNPs_jY22_8-1_dp7miss75miss50.h.hwe | grep -v CHR | awk -F"/" '{print $1" "$2" "$3}' | awk '{print $1" "$2" "$3" "$4" "$5}' | awk '{print $1" "$2" "$3" "$4" "$5" "$4/($3+$4+$5)}' > filteredSNPs_jY22_8-1_dp7miss75miss50_ohz.txt 

#3)
cat filteredSNPs_jY22_8-1_dp7miss75miss50.h.hwe | grep -v CHR | awk '{print $1 " "$2" "$5 " "$6 " "$7}' > filteredSNPs_jY22_8-1_dp7miss75miss50_hwepval 

#4)
grep -v "^##" filteredSNPs_jY22_8-1_dp7miss75miss50.vcf  | cut -f1-3 > filteredSNPs_jY22_8-1_dp7miss75miss50_snpIDs 

```

&nbsp;

#### Step 2: Put heterozygosity in R and make list of passing sites
```{r, message=FALSE, warning=FALSE}
dat <- read.table("~/Documents/crayfish_lab/jaredYseq2022/filteredSNPs_jY22_8-1_dp7miss75miss50_hwepval", header=FALSE)
dat.ohz <- read.table("~/Documents/crayfish_lab/jaredYseq2022/filteredSNPs_jY22_8-1_dp7miss75miss50_ohz.txt", header=FALSE)

dat.SNPids <- read.delim("~/Documents/crayfish_lab/jaredYseq2022/filteredSNPs_jY22_8-1_dp7miss75miss50_snpIDs", header=TRUE)
names(dat.SNPids) <- c("CHROM", "POS", "SNPid")
dat.ohz <- cbind(dat.ohz, dat.SNPids$SNPid)


names(dat.ohz) <- c("CHR", "POS", "AA", "AB", "BB", "Ho", "SNPid")

filtered.ohz <- subset(dat.ohz, AB >= 1 & Ho <= 0.6)

bed_filtered.ohz <- data.frame(CHR = filtered.ohz$CHR, START = filtered.ohz$POS-1, END = filtered.ohz$POS)

snpIDs.filtered.ohz <- data.frame(snp = filtered.ohz$SNPid)

#write.table(snpIDs.filtered.ohz, "~/Documents/crayfish_lab/jaredYseq2022/filteredSNPs_jY22_8-1_dp7miss75miss50_snpIDs.filtered.ohz", sep = "\t", row.names = FALSE, col.names = TRUE, quote = FALSE)

#write.table(bed_filtered.ohz, "~/Documents/crayfish_lab/jaredYseq2022/filteredSNPs_jY22_8-1_dp7miss75miss50_bed_filtered.ohz", sep = "\t", row.names = FALSE, col.names = TRUE, quote = FALSE)

```

&nbsp;

#### Step 3: Filter VCF based on passing sites
```{bash, eval=FALSE}
cd /mnt/home/adamsn23/RedSwampCrayfish/jaredYseq2022/OUT/genotyped
module load GCC/7.3.0-2.30 OpenMPI/3.1.1 VCFtools/0.1.15-Perl-5.28.0

vcftools --vcf filteredSNPs_jY22_8-1_dp7miss75miss50.vcf --snps filteredSNPs_jY22_8-1_dp7miss75miss50_snpIDs.filtered.ohz --recode --recode-INFO-all --stdout > filteredSNPs_jY22_8-1_dp7miss75miss50_hz.vcf

egrep -v "^#" filteredSNPs_jY22_8-1_dp7miss75miss50_hz.vcf | wc -l # 7856
```

&nbsp;

### Remove SNPs with allele balance values >0.6 or <0.4 
#### Step 1: Make list of passing SNPs 
```{r, message=FALSE, warning=FALSE, eval=FALSE}
#### Filter based on allele balance ####

vcf_file <- "~/Documents/crayfish_lab/jaredYseq2022/filteredSNPs_jY22_8-1_dp7miss75miss50_hz.vcf"
seqVCF2GDS(vcf.fn = vcf_file, "~/Documents/crayfish_lab/jaredYseq2022/filteredSNPs_jY22_8-1_dp7miss75miss50_hz.gds", verbose = FALSE)
open_GDS <- seqOpen("~/Documents/crayfish_lab/jaredYseq2022/filteredSNPs_jY22_8-1_dp7miss75miss50_hz.gds", readonly = TRUE)

# - Get sample info
sample_info <- as.data.frame(read.gdsn(index.gdsn(open_GDS, "sample.id")), quote=FALSE)

# data prep
dat <- read.vcfR("~/Documents/crayfish_lab/jaredYseq2022/filteredSNPs_jY22_8-1_dp7miss75miss50_hz.vcf")
tidyDat <- vcfR2tidy(dat)

tmp <- tidyDat$gt %>%  select(ChromKey, POS, Indiv, gt_GT)

tmp2 <- tidyDat$fix %>% select(ChromKey, CHROM, POS, ID)

positionFilter <- left_join(tmp, tmp2, by = c("ChromKey", "POS")) %>%
 mutate(population = substr(Indiv, 1, 5)) %>%
 select(CHROM, POS, ID) %>%
 distinct(ID, .keep_all = TRUE) %>%
 separate(ID, c("tag", "position", NA)) %>%
 mutate(position = as.numeric(position))



# Get allele balance
geno_matrix.012 <-snpgdsGetGeno(open_GDS)
het_matrix <- geno_matrix.012
het_matrix[which(geno_matrix.012 != 1)] <- 0
het_matrix[which(is.na(geno_matrix.012))] <- 0
is.odd <- function(x) x %% 2 != 0
AD <- seqGetData(open_GDS, "annotation/format/AD")
AD <- as.data.frame(AD$data)
ref_c <- AD[,is.odd(seq(1, ncol(AD), 1))]
alt_c <- AD[,!is.odd(seq(1, ncol(AD), 1))]
reads <- ref_c + alt_c
ab1 <- (alt_c/reads)*het_matrix
ab <- colSums(ab1, na.rm = T)/ colSums(het_matrix)
hist(ab)

# Make a list of markers that pass
as_tibble(ab) %>% 
  mutate(AB = ab) %>%
  bind_cols(positionFilter) %>%  ### From SNP count filtering above
  filter(AB > 0.4, 
         AB < 0.6) %>% #position < 141
  drop_na() %>% 
  select(CHROM, POS) #%>% 
  #write.table("~/Documents/crayfish_lab/jaredYseq2022/filteredSNPs_jY22_8-1_dp7miss75miss50_hz_abFilterPass.txt",
   #           quote = FALSE,
    #          row.names = FALSE,
     #         col.names = FALSE)
  
```

&nbsp;

#### Step 2: Filter VCF based on passing SNPs
```{bash, eval=FALSE}
cd /mnt/home/adamsn23/RedSwampCrayfish/jaredYseq2022/OUT/genotyped
module load GCC/7.3.0-2.30 OpenMPI/3.1.1 VCFtools/0.1.15-Perl-5.28.0

vcftools --vcf filteredSNPs_jY22_8-1_dp7miss75miss50_hz.vcf --positions filteredSNPs_jY22_8-1_dp7miss75miss50_hz_abFilterPass.txt --recode --recode-INFO-all --stdout > filteredSNPs_jY22_8-1_dp7miss75miss50_hz.ab.vcf 

egrep -v "^#" filteredSNPs_jY22_8-1_dp7miss75miss50_hz.ab.vcf  | wc -l # 5,307 sites 

```

&nbsp;
&nbsp;

For code to conduct analyses based on juveniles see RSC_ juvenileAnalyses_4ms.Rmd with starting file filteredSNPs_jY22_8-1_dp7miss75miss50_hz.ab.vcf 

&nbsp;

For code to conduct analyses based on berried females, hatchlings, and adults see RSC_multPaternity_4ms.Rmd with starting file filteredSNPs_jY22_8-1_dp7miss75.vcf